### PR TITLE
[dynamo] avoid creating module proxy for fsdp modules

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122876
* #123014
* __->__ #122928
* #123013

FSDP modules overwrites `__new__` so it need `args` and `kwargs`. avoid calling `__new__` without args/kwargs in `nn_module_proxy` function

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Review by williamwen42 — APPROVED** [review]

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -2622,6 +2632,9 @@ def nn_module_proxy(mod):
     if isinstance(mod, torch.fx.GraphModule):
         # Dynamo-generated GM's shouldn't contain user-created GM's
         return mod
+    if fsdp := get_fsdp_module_class():
+        if isinstance(mod, fsdp):
+            return mod
```
Comment (williamwen42) [inline torch/_dynamo/utils.py:2637]: Is it possible to return a "strong" shallow copy of the FSDP module here? If user creates the FSDP module and it reaches here, we may have a reference leak.

Comment (weifengpy) [reply torch/_dynamo/utils.py:2637]: do you refer to `fsdp` or `mod` ? following is my understanding of "strong" shallow copy
```
fsdp = get_fsdp_module_class()
if fsdp is not None and isinstance(mod, fsdp):
    return mod
```

Comment (williamwen42) [reply torch/_dynamo/utils.py:2637]: I’m referring to mod. Retuning mod itself is not a shallow copy - a normal shallow copy is done through copy.copy but this doesn’t copy over everything in a module (e.g. hooks). So I use an approach of calling __new__ in order to create a brand new module with the same underlying data/hooks/etc. Is it possible to create an fsdp module copy using __new__?

Comment (weifengpy) [reply torch/_dynamo/utils.py:2637]: FSDP modules overwrites ``__new__`` so we would need to pass in ``args`` and ``kwargs``

will there be a reference leak if I move the logic to ``output_graph.py`` ? or the memory leak is always there if fsdp is inserted into ``nn_modules_proxies`` ? 
```
nn_modules_proxies = {
      name: nn_module_proxy(mod) for name, mod in self.nn_modules.items() if not instance(mod, fsdp) else name:mod
  }
```

Comment (weifengpy) [reply torch/_dynamo/utils.py:2637]: actually I probably should not let fsdp module reach this step. closing this PR for now. thank @williamwen42 fo review

**Pull request closed without merging**